pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java

```diff
@@ -86,19 +90,29 @@ public int submit(long requestId, QueryPlan queryPlan, long timeoutMs)
           String host = serverInstance.getHostname();
           int servicePort = serverInstance.getQueryServicePort();
           DispatchClient client = getOrCreateDispatchClient(host, servicePort);
-          Worker.QueryResponse response = client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
+          Future<Worker.QueryResponse> response = client.submit(Worker.QueryRequest.newBuilder().setStagePlan(
                   QueryPlanSerDeUtils.serialize(constructDistributedStagePlan(queryPlan, stageId, serverInstance)))
               .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_ID, String.valueOf(requestId))
               .putMetadata(QueryConfig.KEY_OF_BROKER_REQUEST_TIMEOUT_MS, String.valueOf(timeoutMs)).build());
-          if (response.containsMetadata(QueryConfig.KEY_OF_SERVER_RESPONSE_STATUS_ERROR)) {
-            throw new RuntimeException(
-                String.format("Unable to execute query plan at stage %s on server %s: ERROR: %s", stageId,
-                    serverInstance, response));
-          }
+          querySubmitResponse.add(response);
+        }
+      }
+    }
+    while (System.currentTimeMillis() < submissionDeadlineMs) {
+      for (Future<Worker.QueryResponse> future : querySubmitResponse) {
+        if (!future.isDone()) {
+          break;
+          Thread.sleep(100);
+        }
+      }
+      for (Future<Worker.QueryResponse> future : querySubmitResponse) {
+        if (future.get().containsMetadata(QueryConfig.KEY_OF_SERVER_RESPONSE_STATUS_ERROR)) {
```

Comment:
Does this check for actual query response from server or the submission acknowledgement ?

Reply:
no it doesn't